### ISSUE TITLE
Feature: WebTorrent Integration

### ISSUE THREAD

**Issue by mikeal** [brave/browser-laptop#3256]
It would be great if magnet links downloaded right in the browser. The code to integrate is already quite accessible via [WebTorrent](https://github.com/feross/webtorrent).

If you aren't familiar WebTorrent:
- Works in-browser over WebRTC
- Works in Node.js over WebRTC _and_ standard Bittorrent protocols and will even act as a bridge node.
- The [desktop client](https://github.com/feross/webtorrent-desktop) is written for Electron so if you're looking for a place to start it's pretty accessible.

/cc @feross

**Comment by feross** [comment]
Nice idea, @mikeal. This would definitely set Brave apart from other browsers. Built-in torrent support would be really cool, and not that hard to implement with `webtorrent`.

Here's how this could work:

- When a magnet link or `.torrent` file is clicked, initialize a WebTorrent instance add the link.
  ```js
  var client = new WebTorrent()
  var torrent = client.add(magnetLink)
  ```

- Once the metadata is loaded (usually takes ~2 seconds), show the files in the torrent.
  ```js
  torrent.once('metadata', function () {
    torrent.files.forEach(function (file) {
      console.log(file.name, file.length) // render the files in the UI
    })
  }
  ```

- When a file is selected, stream that specific file only.
  ```js
  file.createReadStream() // get Node.js style Readable stream with file data, render it
  ```
  
- When the user navigates away, destroy the torrent and (optionally) the WebTorrent client to close all connections and reclaim all resources.
  ```js
  torrent.destroy()
  client.destroy()
  ```

It also probably makes sense to warn the user before starting a torrent on their behalf so they can make an informed decision about whether they wish to join the swarm.

**Comment by feross** [comment]
If you don't want to implement any custom UI, you can actually hack this together really quickly using the `torrent.createServer()` feature in WebTorrent.

```js
var server = torrent.createServer() // create HTTP server for this torrent
server.listen(PORT, HOSTNAME, function () { // start the server listening to a port
  // Navigate the user to http://<hostname>:<port>/ to see a list of files, with links to each file

  // Access individual files at http://<hostname>:<port>/<index> where index is the index
  // in the torrent.files array
})
```

**Comment by TheAlphaNerd** [comment]
how hard would it be to make it pipe through a VPN?

**Comment by feross** [comment]
@TheAlphaNerd There is an open PR on WebTorrent to add SOCKS proxy support. Not sure what's required for VPN support.

**Comment by TheAlphaNerd** [comment]
@feross I think we will be at some of the same events in the next couple months. maybe we can find time to hack on it 😇

**Comment by feross** [comment]
@TheAlphaNerd That would be fun!

**Comment by DiegoRBaquero** [comment]
This would be awesome! @TheAlphaNerd If the OS is using a VPN everything WebTorrent does should be behind it already. I don't really know if the browsers SOCKS setting would put WebRTC connections behind it, @feross ?

**Comment by luixxiul** [comment]
Would it be possible to use this feature while IP address leakage via WebRTC is disabled with the fingerprinting protection? / cc: @diracdeltas 

#260

**Comment by retrohacker** [comment]
We could combine this with serving static files from a torrent over a custom protocol:

https://github.com/retrohacker/peerweb/blob/master/index.js#L65-L134

**Comment by retrohacker** [comment]
Note: the http fallback there is a nasty hack, when https://github.com/electron/electron/issues/4303 lands, a redirect will solve that.

**Comment by feross** [comment]
@luixxiul Yes, WebTorrent should work just fine with the fingerprinting detection in Brave. With that setting enabled, WebRTC just finds detects fewer ip:port candidates. Specifically, it would just use the ip:port of the default route (public IP, or if a VPN is in use, then the VPN's IP). Normally WebRTC finds additional IPs, including the local IP address, but that's just to increase the chances of a successful connection (more possible routes) and to improve the performance of the connection (local routes will be faster than public routes).

tldr: WebTorrent's WebRTC functionality will operate fine with the reduced list of IPs that Brave's fingerprinting mode returns.

**Comment by feross** [comment]
Hey torrent fans! I'm working on adding WebTorrent to Brave with @dcposch over the next couple weeks. Here's the design doc that explains our current plan: https://docs.google.com/document/d/1d_NXJp05gGcy8kYYDcntIRPQaDrvU3CUF-QjEpf0PLI/edit?usp=sharing

Feedback and comments welcome!

**Comment by feross** [comment]
Initial PR to add WebTorrent support is out! https://github.com/brave/browser-laptop/pull/5342

🎉 🎉 🎉

**Comment by luixxiul** [comment]
Can we close this?

**Comment by luixxiul** [comment]
Would someone please specify the steps for QA? They must have been specified somewhere in a PR, but it should be clarified here for reference as the QA work should start from the issue, not a PR.

Thanks in advance!

**Comment by feross** [comment]
@luixxiul There's a design doc here that will probably be helpful: https://docs.google.com/document/d/1d_NXJp05gGcy8kYYDcntIRPQaDrvU3CUF-QjEpf0PLI/edit?usp=sharing

Support the following in the Brave desktop app:
 - Handle navigation to magnet links and torrent files. For example:
    - magnet:?xt=urn:btih:6a9759bffd5c0af65319979fb7832189f4f3c35d&dn=sintel
 - Show torrent content within Brave, without the need for an external torrent app.

**Comment by luixxiul** [comment]
@feross thanks, I didn't know `.torrent` files were planned but not supported yet. I was almost creating an issue for it.

**Comment by srirambv** [comment]
QA Steps 
1. Open any magnet torrent link 
2. Torrent should start downloading

QA possible once #5981 is fixed
